### PR TITLE
DVX-79 npl-integrations improve README

### PR DESCRIPTION
**Pull request by jean-noumena**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,63 +1,163 @@
 # npl-integrations
 
-## Pre-requisites
-Make sure your `hosts` file includes keycloak at 127.0.0.1:
+The npl-integrations repo contains a sample project that demonstrates how to integrate with the NPL engine using
```
Comment (ari-noumena) [inline README.md:3]: Maybe 

```suggestion
The npl-integrations repo contains a sample project that demonstrates how to integrate with [Noumena's Operating Engine](https://documentation.noumenadigital.com/engine/applications/engine/) using
```

File: README.md [hunk 2]
```diff
@@ -1,63 +1,163 @@
...
+different programming languages and frameworks.
+It includes a Python listener service, a typescript React webapp, and a python Streamlit UI.
```
Comment (ari-noumena) [inline README.md:5]:
```suggestion
It includes a Python listener service, a typescript React web app, and a Python Streamlit UI.
```

File: README.md [hunk 3]
```diff
@@ -1,63 +1,163 @@
...
 
 ## Building and running with docker locally
-1. Run `mvn clean install` to build and generate NPL-api and clients.
-2. Run `docker compose up --build -d` to ensure python and frontend are also build, then create and start containers.
 
-URLs:
-* Swagger UI of Engine APIs: http://localhost:12000/swagger-ui/
-* Keycloak admin console: http://localhost:11000
-* Webapp: http://localhost:8090
+The first deployment possibility consists of running the NPL code and all services locally.
```
Comment (ari-noumena) [inline README.md:12]: I don't think there's really any conceptual difference between deploying to a prod environment and deploying locally, so I would rephrase this to just say that we deploy to an Engine container, which you _can_ run locally.

File: README.md [hunk 4]
```diff
@@ -1,63 +1,163 @@
...
+
+For OS X and linux systems, make sure the `/etc/hosts` file includes the line `127.0.0.1 keycloak`
+
+![img.png](docs/img.png)
+
+If the line is not present, follow the steps below to add it:
+
+1. Use the `sudo vim /etc/hosts` command to open the file in a text editor. Enter your password when prompted for this admin command.
+2. Navigate to the end of the file, press `i` to enter insert mode and add the line `127.0.0.1 keycloak`.
+3. Save by pressing `Esc`, then type `:wq` and press `Enter`.
```
Comment (ari-noumena) [inline README.md:28]: Or

```
echo "127.0.0.1 keycloak" | sudo tee -a /etc/hosts
```

File: README.md [hunk 5]
```diff
@@ -1,63 +1,163 @@
...
+
+### Build & run
+
+To build and run the project, follow the steps below:
+
+1. Run `mvn clean install` to build and generate NPL-API and clients.
+2. Run `docker compose up --build -d` to ensure the python listener service, the python Streamlit UI and the typescript React frontend are build in addition to engine and engine dependencies, then create and start containers.
```
Comment (ari-noumena) [inline README.md:35]:
```suggestion
2. Run `docker compose up --build -d` to ensure the Python listener service, the Python Streamlit UI and the Typescript React frontend are built in addition to the engine and its dependencies, then create and start the containers.
```

File: README.md [hunk 6]
```diff
@@ -1,63 +1,163 @@
...
+
+### Service endpoints
+
+Once the project is running, services run behind the following URLs:
```
Comment (ari-noumena) [inline README.md:39]:
```suggestion
Once the project is running, services can be accessed with the following URLs:
```

File: README.md [hunk 7]
```diff
@@ -1,63 +1,163 @@
...
+
+The second deployment possibility consists of running the NPL code on Noumena Cloud and local services.
+Noumena offers a cloud-based environment for running NPL code, which can be accessed at [portal.noumena.cloud](https://portal.noumena.cloud).
+
+This setup includes running a python listener, python Streamlit UI & typescript React webapp locally, and Noumena Engine on Noumena Cloud.
+
+### NPL and keycloak
+
+In this setup, NPL code runs on Noumena Cloud. Supporting services of the NPL Engine are deployed alongside the NPL Engine.
+Supporting services include Keycloak for authentication and authorization and databases.
```
Comment (ari-noumena) [inline README.md:59]:
```suggestion
Supporting services include Keycloak for authentication and authorization, and databases.
```

File: README.md [hunk 8]
```diff
@@ -1,63 +1,163 @@
...
+1. Install the NPL CLI by running `make download-cli` in the root directory.
+2. Add the following variables to your shell to configure the NPL CLI and restart your terminal
+
+    ```
+    export NC_BASE_URL=https://portal.noumena.cloud
+    export NC_EMAIL=your.email@your.domain
+    export NC_ENV=PROD
+    ```
+
+3. Run `make create-app` to create the application with name defined in Makefile.
```
Comment (ari-noumena) [inline README.md:72]:
```suggestion
3. Run `make create-app` to create the application with the name defined in Makefile.
```

File: README.md [hunk 9]
```diff
@@ -1,63 +1,163 @@
...
+4. Run `make iam` to provision keycloak on the created application with terraform.
+5. Run `make clear-deploy` to clear pre-existing packages in the app and upload current NPL and migration sources.
```
Comment (ari-noumena) [inline README.md:74]:
```suggestion
5. Run `make clear-deploy` to clear pre-existing packages in the app and upload the current NPL and migration sources.
```

File: README.md [hunk 10]
```diff
@@ -1,63 +1,163 @@
...
+
+#### Option 2: Using the maven plugin
```
Comment (ari-noumena) [inline README.md:76]: ...NPL-dev for IntelliJ, right? Though this is a different option: https://documentation.noumenadigital.com/tools/build-tools/Engine-Maven-Plugin/#deploy-source

File: README.md [hunk 11]
```diff
@@ -1,63 +1,163 @@
...
+The `./python-listener` folder contains a python service interacting with the configured engine.
+In this setup, the python listener service runs locally and connects to the engine on Noumena Cloud.
 
 #### Setup
 
-1. set up & activate your Python venv
-2. run `mvn install` to generate the python client
-3. run `cd python && pip install -r requirements.txt` 
+1. In the root directory, set up & activate your Python venv
+2. run `mvn install` to generate the python client from the NPL code
```
Comment (ari-noumena) [inline README.md:104]:
```suggestion
2. run `mvn install` to generate the Python client from the NPL code
```

File: README.md [hunk 12]
```diff
@@ -1,63 +1,163 @@
...
-
-and in another terminal
-
-run `cd python && python notification.py`
+From the root directory, run `cd python-listener && python main.py`
 
 ### Webapp
 
-The './webapp' folder contains a typescript frontend service interacting with the configured engine.
+The `./webapp` folder contains a typescript frontend service interacting with the configured engine.
```
Comment (ari-noumena) [inline README.md:116]:
```suggestion
The `./webapp` folder contains a Typescript frontend service for interacting with the configured Engine.
```

Comment (ari-noumena) [reply README.md:116]: I guess in general just make sure `Python`, `Engine`, and `Typescipt` are proper nouns everywhere

File: README.md [hunk 13]
```diff
@@ -1,63 +1,163 @@
...
+| Swagger UI of Engine APIs | `https://engine-$NC_ORG_NAME-$NC_APP_NAME.noumena.cloud/swagger-ui/` |
+| Keycloak admin console    | `https://keycloak-$NC_ORG_NAME-$NC_APP_NAME.noumena.cloud`           |
+| Webapp                    | http://localhost:5173                                                |
+| Streamlit UI              | http://localhost:8501                                                |
+
+## Next steps
+
+You now know everything about NPL deployment to start developing your own application on top of the NPL engine in your preferred language.
+
+For more information on the NPL engine, please refer to
```
Comment (ari-noumena) [inline README.md:162]:
```suggestion
For more information about the Engine, please refer to
```

File: webapp/README.md [hunk 14]
```diff
@@ -1 +1,61 @@
-# IOU UI
+# IOU Typescript Webapp
+
+This project is a webapp that demonstrates how to integrate with the NPL engine using the IOU extension.
+It is built using Typescript and React. It allows users to create, read and perform actions of IOUs.
+
+## Maven generator
+
+The selected maven generator for typescript is 'typescript-axios'.
```
Comment (ari-noumena) [inline webapp/README.md:8]:
```suggestion
The selected Maven generator for typescript is 'typescript-axios'.
```

I think Maven needs to be a proper noun everywhere as well

Comment (ari-noumena) [inline webapp/README.md:8]: Maybe a link rather than ''s?

File: webapp/README.md [hunk 15]
```diff
@@ -1 +1,61 @@
...
+It solves a few compilation and setup errors.
```
Comment (ari-noumena) [inline webapp/README.md:9]: Does this need to be mentioned?

File: webapp/README.md [hunk 16]
```diff
@@ -1 +1,61 @@
...
+
+## Authentication in the API service
+
+The authentication is done through the keycloak service.
+The keycloak service is configured in the `src/KeycloakProvider.tsx` file and is used to authenticate the user and get the access token.
+The access token is then used to authenticate the API service in the `src/services/BaseService.tsx` and attached as header to http requests.
```
Comment (ari-noumena) [inline webapp/README.md:15]:
```suggestion
The access token is then used to authenticate the API service in `src/services/BaseService.tsx` and attached as a header to HTTP requests.
```

File: webapp/README.md [hunk 17]
```diff
@@ -1 +1,61 @@
...
+
+```typescript
+    private withAuthorizationHeader = () => {
+        return { headers: { Authorization: `Bearer ${this.keycloak.token}` } }
+    }
+```
+
+## API service
+
+The API service is implemented in the `src/services/BaseService.tsx` file and is used to interact with the NPL engine.
```
Comment (ari-noumena) [inline webapp/README.md:25]:
```suggestion
The API service is implemented in `src/services/BaseService.tsx` and is used to interact with the Engine.
```

File: webapp/README.md [hunk 18]
```diff
@@ -1 +1,61 @@
...
+It relies on the generated sources from the NPL engine to make requests to the engine.
```
Comment (ari-noumena) [inline webapp/README.md:26]: >from the NPL engine

From the NPL, presumably?

File: README.md (hunk 10 above)
Comment (jean-noumena) [reply README.md:76]: https://noumenadigital.atlassian.net/browse/DVX-80

File: README.md [hunk 19]
```diff
@@ -1,63 +1,161 @@
...
 
 ## Building and running with docker locally
-1. Run `mvn clean install` to build and generate NPL-api and clients.
-2. Run `docker compose up --build -d` to ensure python and frontend are also build, then create and start containers.
 
-URLs:
-* Swagger UI of Engine APIs: http://localhost:12000/swagger-ui/
-* Keycloak admin console: http://localhost:11000
-* Webapp: http://localhost:8090
+The project can be built and run locally using docker-compose.
```
Comment (ari-noumena) [inline README.md:12]:
```suggestion
The project can be built and run locally using Docker Compose.
```

To avoid any confusion with the deprecated binary (or add .yml if you mean the file itself).

Comment (ari-noumena) [reply README.md:12]: Also below

File: README.md [hunk 20]
```diff
@@ -1,63 +1,161 @@
...
+
+The npl-integrations project can be locally built and all containers run locally using docker-compose.
+
+### Pre-requisites
+
+For OS X and linux systems, make sure the `/etc/hosts` file includes the line `127.0.0.1 keycloak`
+
+![img.png](docs/img.png)
+
+If the line is not present, follow the steps below to add it by running `echo "127.0.0.1 keycloak" | sudo tee -a /etc/hosts` in the terminal.
```
Comment (ari-noumena) [inline README.md:24]: There are no steps now :D

File: README.md [hunk 21]
```diff
@@ -1,63 +1,161 @@
...
-* Webapp: http://localhost:8090
+The project can be built and run locally using docker-compose.
+To run the Noumena Engine, the corresponding docker image is required, which is available under license. 
+Please contact Noumena Digital for more information at [info@noumenadigital.com](mailto:info@noumenadigital.com)
+
+The npl-integrations project can be locally built and all containers run locally using docker-compose.
+
+### Pre-requisites
+
+For OS X and linux systems, make sure the `/etc/hosts` file includes the line `127.0.0.1 keycloak`
```
Comment (ari-noumena) [inline README.md:20]: This is presumably also a requirement on Windows, though that is where the file lives on MacOS and Linux.

File: README.md [hunk 22]
```diff
@@ -1,63 +1,161 @@
...
+![img.png](docs/img.png)
+
+If the line is not present, follow the steps below to add it by running `echo "127.0.0.1 keycloak" | sudo tee -a /etc/hosts` in the terminal.
+
+### Build & run
+
+To build and run the project, follow the steps below:
+
+1. Run `mvn clean install` to build and generate NPL-API and clients.
+2. Run `docker compose up --build -d` to ensure the Python listener service, the Python Streamlit UI and the Typescript React frontend are built in addition to the engine and its dependencies, then create and start the containers.
```
Comment (ari-noumena) [inline README.md:31]:
```suggestion
2. Run `docker compose up --build -d` to ensure the Python listener service, the Python Streamlit UI and the Typescript React frontend are built in addition to the Engine and its dependencies, then create and start the containers.
```

Double check the proper nouns :D -- (sorry, this really is a nit, though Engine is probably one of the ones that actually do matter, a little bit)

**Review by ari-noumena — DISMISSED** [review]
These are some nice docs and I think we should probably collate more newbie-friendly deployment docs (with all the different options) in the docs proper in the future, and this would be a great baseline for that.

File: README.md [hunk 23]
```diff
@@ -1,63 +1,166 @@
...
+
+### Service endpoints
+
+Once the project is running, services can be accessed with the following URLs:
+
+| Service                   | URL                                |
+|---------------------------|------------------------------------|
+| Engine APIs               | http://localhost:12000/            |
+| Swagger UI of Engine APIs | http://localhost:12000/swagger-ui/ |
+| Keycloak admin console    | http://localhost:11000             |
```
Comment (pierre-noumena) [inline README.md:46]: I believe using http://keycloak:11000 instead of http://localhost:11000 removes redirect errors when you visit that page and klick to access the actual admin console.

**Review by pierre-noumena — APPROVED** [review]
